File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/I.Любители_конференций/solution.py
import sys


def solution(n: int, universities_id: list[int], k: int) -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    """
    hashmap: dict[int, int] = dict()
    for university_id in universities_id:
        hashmap[university_id] = hashmap.get(university_id, 0) + 1
    universities_id_distinct: list[int] = sorted(hashmap.keys(), reverse=True, key=lambda university_id: (hashmap[university_id], -university_id))
    for idx in range(min(len(universities_id_distinct), k)):
        if idx:
            sys.stdout.write(' ')
        sys.stdout.write(str(universities_id_distinct[idx]))
    sys.stdout.write('\n')


if __name__ == '__main__':
    n: int = int(sys.stdin.readline().rstrip())
    universities_id: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    k: int = int(sys.stdin.readline().rstrip())
    solution(n, universities_id, k)

File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/I.Любители_конференций/solution.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <algorithm>


void solution(int& n, std::vector<int>& universitiesId, int& k) {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    */
    std::unordered_map<int, int> hashmap;
    for (int& universityId : universitiesId) {
        ++hashmap[universityId];
    }
    std::vector<int> distinctUniversitiesId;
    for (auto& pair : hashmap) {
        distinctUniversitiesId.push_back(pair.first);
    }
    std::sort(
        distinctUniversitiesId.begin(),
        distinctUniversitiesId.end(),
        [&hashmap](const int& x, const int& y) {
            if (hashmap[x] != hashmap[y]) {
                return hashmap[x] > hashmap[y];
            }
            return -x > -y;
        }
    );
    int dn = distinctUniversitiesId.size();
    for (int idx = 0; idx < std::min(dn, k); ++idx) {
        if (idx) {
            std::cout << " ";
        }
        std::cout << distinctUniversitiesId[idx];
    }
    std::cout << "\n";
}


int main() {
    int n;
    std::cin >> n;
    std::vector<int> universitiesId;
    for (int i = 0; i < n; ++i) {
        int universityId;
        std::cin >> universityId;
        universitiesId.push_back(universityId);
    }
    int k;
    std::cin >> k;
    solution(n, universitiesId, k);
}